### PR TITLE
Add SDL_RenderWindowToLogical() and SDL_RenderLogicalToWindow to Render API

### LINKED ISSUE #1397: SDL_GetMouseState() seems to give raw mouse position when using SDL_RenderSetLogicalSize()
# This bug report was migrated from our old Bugzilla tracker.

These attachments are available in the static archive:

* [shows bug (mousestate.c, text/x-csrc, 2014-03-11 01:54:46 +0000, 852 bytes)](https://bugzilla.libsdl.org/attachment.cgi?id=1589)

**Reported in version:** HG 2.1
**Reported for operating system, platform:** Linux, x86_64

# Comments on the original bug report:

On 2014-03-11 01:41:25 +0000,  wrote:

> As summary. MOUSEBUTTONDOWN events report the position as expected (in logical screen coordinates). The window is resizable. Probably also happens with fixed-size windows and a logical size that is a different aspect ratio than the window itself.

On 2014-03-11 01:54:46 +0000,  wrote:

> Created attachment 1589
> shows bug
> 
> Here is an example program that shows the bug. On my machine, clicking in the corners of the white rectangle:
>
> ```
> $ ./mousestate 
> MOUSEBUTTONDOWN (1, 2)
> GetMouseState (102, 3)
> 
> MOUSEBUTTONDOWN (398, 398)
> GetMouseState (697, 598)
> ```

On 2014-05-10 15:13:51 +0000, Leonardo wrote:

> This is true. The problem is that GetMouseState() consults the x/y directly whereas when reported via events, the renderer changes all the mouse events first with an event watcher before they go to the queue.
> 
> I don't see this being fixed easily as it would require the GetMouseState function to consult the renderer.
> 
> Actually this can come in handy if you ever need the real mouse position as there is no other way of knowing it. If you require your mouse coordinates to be adjusted to the logical size, you can create a event watcher like this:
> 
> ```c
> int MouseWatcher(void* userdata, SDL_Event* event) {
>   MouseState * m = (MouseState*) userdata;
>   switch(event->type) {
>     case SDL_MOUSEBUTTONDOWN:
>     case SDL_MOUSEBUTTONUP:
>       m->x = event->button.x;
>       m->y = event->button.y;
>       break;
>     case SDL_MOUSEMOTION: 
>       m->x = event->motion.x;
>       m->y = event->motion.y;
>       break;
>    default:
>       break;
>   }
>   return 0;
> }
> ```
>
> And then add it: 
>
> ```c 
> {
>   SDL_AddEventWatch(MouseWatcher, mouseState);
> }
> ```
>
> Note that you'll need at least 2.0.3 because of bug # 2414

On 2014-07-09 07:09:03 +0000, Sylvain wrote:

> Hey, I am not the original poster but I faced the same issue.
> Just some thought:
> 
> CreateRenderer assume a Window.
> But could we created several Renderer for the Same Window?
> 
> If so, converting an "event" (mouse position) from Window Frame to Renderer Frame is king of buggy, we could only convert to 1 specific Renderer Frame (or to the only renderer if only 1 one exists).
> 
> If no, maybe we could add a function to get the renderer of a Window : 
> 
> SDL_renderer * SDL_GetRendererOfWindow(SDL_window*)
> 
> Also. Maybe the code of "SDL_RendererEventWatch" could be factorized to have a function like :
> 
> "convert_event_in_window_coordinates_to_renderer_coordinates"
> 
> We would be used in SDL_RendererEventWatch, and also exported so that users who call SDL_GetMouseState can convert their event to LogicalRendererSize.
> 
> Cheers,
> 
> Sylvain

On 2015-01-09 20:55:05 +0000,  wrote:

> It seems to me the resolution of this is just to add Leonardo's comments to the docs/wiki. Right now it doesn't say that SDL_GetMouseState() is the only way to get the real mouse position. In light of that, it makes sense to just get the computed position from the motion or button events.

On 2016-06-09 03:02:31 +0000, Arne Döring wrote:

> I would like to ask, if there is anything done on this bug? I found out about this bug, too, and because the logical size didn't even do a recalculation of the mouse coordinates, I decided, to not use the logical size anymore. I really feels like a not well thought through feature, that I wouldn't recommend to anyone who uses SDL.
> 
> But the not make this post sound too negative. Thanks for the great work to have SDL2 actually done. I enjoy working with it.

On 2019-10-17 11:45:16 +0000, Sylvain wrote:

> It's reasonable to say that querying the state returns raw values,!
> We can revisit this for SDL 2.1
> (see also bug 4028)

### PR DESCRIPTION
**Pull request by johnblat64**
## Description

Adds a function that when given real coordinates of some point in a renderer's window, will return the logical coordinates of a point in relation the to the renderer.

Useful for when the renderer is scaled and has a set logical size.  

## Existing Issue(s)
resolves #1397

### CODE REVIEW

File: include/SDL_render.h [hunk 1]
```diff
@@ -979,6 +979,25 @@ extern DECLSPEC int SDLCALL SDL_RenderSetScale(SDL_Renderer * renderer,
...
                                                float *scaleX, float *scaleY);
 
+/**
+ * Get logical coordinates of point in renderer when given real coordinates of point in window.
+ * Logical coordinates will differ from real coordinates when render is scaled and logical renderer size set
+ * 
+ * \param renderer the renderer from which the logical coordinates should be calcualted
+ * \param realX the real X coordinate in the window
+ * \param realY the real Y coordinate in the window
+ * \param logicalX the pointer filled with the logical x cooridnate
```
Comment (slouken) [inline include/SDL_render.h:989]: typo

File: src/render/SDL_render.c [hunk 2]
```diff
@@ -2491,6 +2491,42 @@ SDL_RenderGetScale(SDL_Renderer * renderer, float *scaleX, float *scaleY)
...
+    /* Compare with midpoints because a renderer that is logically sized will be centered and have a border
+     if the window aspect ratio differs from the renderer's.
+    */
+    window_renderer_midpoint_delta_x = window_midpoint_x - renderer_midpoint_x;
+    window_renderer_midpoint_delta_y = window_midpoint_y - renderer_midpoint_y;
+
+    renderer_real_x = realX - window_renderer_midpoint_delta_x; 
+    renderer_real_y = realY - window_renderer_midpoint_delta_y;
+
+    *logicalX = (float)renderer_real_x / (float)renderer_real_w * (float)renderer->logical_w;
```
Comment (slouken) [inline src/render/SDL_render.c:2526]: The logical coordinate should probably be float, since we can render with floating point precision.

**Comment by johnblat64** [comment]
Resolved comments. Let me know if there's anything else!

**Comment by slouken** [comment]
Thanks, your changes have been merged as part of commit https://github.com/libsdl-org/SDL/commit/0d98793693fa2e428a578efbf3e761aea2d0b4dd.

**Pull request closed without merging**